File: 22. Generate Parentheses/main.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include <thread>
#include <algorithm>

using namespace std;

class Solution {
    public:
        vector<string> generateParenthesis(int n) {
        vector<string> res;
        generateParenthesisDFS(n, n, "", res);
        return res;
    }
    
    // left, right: remaining  #"(" or ")"
    void generateParenthesisDFS(int left, int right, string temp, vector<string>& res) {
        if (left > right) return ;
        if (left == 0 && right == 0) res.push_back(temp);
        else {
            if (left >0) generateParenthesisDFS(left-1, right, temp+"(", res);
            if (right>0) generateParenthesisDFS(left, right-1, temp+")", res);
        }
    }
};

int main(int argc, char** argv) {
    int n = 6;
    vector<string> res = Solution().generateParenthesis(n);
    for(auto& i : res) {
        cout << i << endl;
    }
    return 0;
}

File: 22. Generate Parentheses/src/main/Main.java
package main;

import java.util.ArrayList;
import java.util.List;

class Solution {
    
	public List<String> generateParenthesis(int n) {
    	List<String> res = new ArrayList<>();
    	util("", res, 0, 0, n);
    	return res;
    }
    
    void util(String s, List<String> res, int open, int close, int n) {
    	if(s.length() == n*2) {
    		res.add(s);
    		return;
    	}
    	//open is count of '(' and close ')'
    	if(open < n)
    		util(s+"(", res, open+1, close, n);
    	if(close < open)
    		util(s+")", res, open, close+1, n);    		
    }
}

public class Main {
	public static void main(String[] args) {
		int n = 3;
		System.out.println(new Solution().generateParenthesis(n));
	}
}